imagecraft/pack/grubutil.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2025 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""GRUB utils."""
+
+import subprocess
+from pathlib import Path
+
+from craft_cli import emit
+from craft_platforms import DebianArchitecture
+
+from imagecraft import errors
+from imagecraft.pack.chroot import Chroot, Mount
+from imagecraft.pack.image import Image
+from imagecraft.subprocesses import run
+
+_ARCH_TO_GRUB_EFI_TARGET: dict[str, str] = {
+    DebianArchitecture.AMD64: "x86_64-efi",
+    DebianArchitecture.ARM64: "arm64-efi",
+    DebianArchitecture.ARMHF: "arm-efi",
+}
+
+
+def _grub_install(grub_target: str, loop_dev: str) -> None:
+    """Install grub in the image.
+
+    :param grub_target: target platform to install grub for.
+    :param loop_dev: loop device to install grub on
+    """
+    check_grub_install = ["grub-install", "-V"]
+    grub_install_command = [
+        "grub-install",
+        loop_dev,
+        "--boot-directory=/boot",
+        "--efi-directory=/boot/efi",
+        f"--target={grub_target}",
+        "--uefi-secure-boot",
+        "--no-nvram",
+    ]
+
+    update_grub_command = [
+        "update-grub",
+    ]
+
+    # Divert os-probe to avoid writing wrong output in grub.cfg
+    os_prober = "/etc/grub.d/30_os-prober"
+    divert_base_command = "dpkg-divert"
+
+    divert_common_args = [
+        "--local",
+        "--divert",
+        os_prober + ".dpkg-divert",
+        "--rename",
+        os_prober,
+    ]
+
+    divert_os_prober_command = [divert_base_command, *list(divert_common_args)]
+
+    undivert_os_prober_command = [
+        divert_base_command,
+        "--remove",
+        *divert_common_args,
+    ]
+
+    # Check if grub-install is available, otherwise skip the installation without error
+    try:
+        run(*check_grub_install)
+    except FileNotFoundError:
+        emit.message("Skipping GRUB installation because grub-install is not available")
```

Comment:
I would have thought that if we reach this point, the image creation is doomed. The error sounds like this is an optional step that can be skipped ?

Reply:
We only want to try to install GRUB if some conditions are met, otherwise we conclude that the user is trying to build another kind of image (maybe with bootloader provided in another way) so it is fine to do nothing.
This piece of code is the last check, after that a failure will really fail the build.